### PR TITLE
First stab: extracting the checkpoint delta from the indexed split.

### PR DESCRIPTION
**Pull request by fulmicoton**
Refactoring to prepare partitioning.

- NewSplit and Replace messages are merged.
- Delta is extracted from the IndexedSplitmessage
- Added a CheckpointDelta / SourceCheckpointDelta distinction.
- Moved source_id into the checkpoint delta.

### CODE REVIEW

File: quickwit-indexing/src/actors/indexer.rs [hunk 1]
```diff
@@ -239,11 +257,16 @@ impl IndexerState {
     }
 }
 
+struct IndexingWorkbench {
```
Comment (guilload) [inline quickwit-indexing/src/actors/indexer.rs:260]: What an indexing workbench is is not exactly intuitive. Might be worth adding a comment in the follow-up PR that handles multiple splits in the workbench.

File: quickwit-indexing/src/actors/uploader.rs [hunk 2]
```diff
@@ -216,32 +216,24 @@ fn create_split_metadata(split: &PackagedSplit, footer_offsets: Range<u64>) -> S
...
-        })
+    SplitUpdate {
+        index_id,
+        new_splits: packaged_splits_and_metadatas
+            .into_iter()
+            .map(|split_and_meta| split_and_meta.1)
+            .collect_vec(),
+        replaced_split_ids: Vec::from_iter(replaced_split_ids),
+        checkpoint_delta_opt,
+        //         split_date_of_birth: packaged_split.split_date_of_birth,
```
Comment (guilload) [inline quickwit-indexing/src/actors/uploader.rs:235]: You're fixing this in the next PR? One `Instant` per split? Created when the first doc is added?

File: quickwit-indexing/src/models/publisher_message.rs [hunk 3]
```diff
@@ -20,55 +20,32 @@
...
 
-impl fmt::Debug for PublishNewSplit {
+impl fmt::Debug for SplitUpdate {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("PublishNewSplit")
+        let new_split_ids: Vec<&str> = self
+            .new_splits
+            .iter()
+            .map(|split| split.split_id())
+            .collect();
```
Comment (guilload) [inline quickwit-indexing/src/models/publisher_message.rs:40]: nit: Straight into a String with itertools `.join(", ")`?

File: quickwit-indexing/src/models/raw_doc_batch.rs [hunk 4]
```diff
@@ -19,12 +19,27 @@
...
+impl RawDocBatch {
+    pub fn new(docs: Vec<String>, checkpoint_delta: SourceCheckpointDelta) -> Self {
+        RawDocBatch {
+            docs,
+            checkpoint_delta,
+        }
+    }
+}
+
+impl Default for RawDocBatch {
```
Comment (guilload) [inline quickwit-indexing/src/models/raw_doc_batch.rs:39]: Can't we derive `Default` for `RawDocBatch` if `Vec` and `SourceCheckpointDelta` do?

**Review by guilload — APPROVED** [review]

File: quickwit-indexing/src/actors/indexer.rs [hunk 5]
```diff
@@ -123,34 +129,42 @@ impl IndexerState {
             ctx.kill_switch().clone(),
         )?;
         info!(split_id = %indexed_split.split_id, "new-split");
-        Ok(indexed_split)
+        let workbench = IndexingWorkbench {
+            checkpoint_delta: IndexCheckpointDelta {
+                source_id: self.source_id.clone(),
+                source_delta: SourceCheckpointDelta::default(),
+            },
+            indexed_split,
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit-indexing/src/actors/indexer.rs:137]: yes

File: quickwit-indexing/src/actors/uploader.rs (hunk 2 above)
Comment (fulmicoton) [reply quickwit-indexing/src/actors/uploader.rs:235]: This PR ditch the split date of birth for a better date that is emitted by the source.

**Pull request merged**